bake/hcl_test.go

```diff
@@ -1645,6 +1646,692 @@ func TestHCLIndexOfFunc(t *testing.T) {
+	})
+
+	// this is implied/exercised in other tests, but repeated for completeness
+	t.Run("JSON override ignored if var is untyped", func(t *testing.T) {
+		dt := []byte(`
+            variable "FOO" {
+                default = [1, 2]
+            }
+            
+            target "default" {
+                args = {
+                    foo = length(FOO)
+                }
+            }`)
+		t.Setenv("FOO_JSON", "[3,4]")
+		_, err := ParseFile(dt, "docker-bake.hcl")
+		require.ErrorContains(t, err, "unsupported type")
+	})
+
+	t.Run("override-ish variable has regular CSV override", func(t *testing.T) {
+		dt := []byte(`
+            variable "FOO_JSON" {
+                type = list(number)
+            }
+
+            target "default" {
+                args = {
+                    foo = length(FOO_JSON)
+                }
+            }`)
+		// despite the name, it's still CSV
+		t.Setenv("FOO_JSON", "10,11,12")
+		c, err := ParseFile(dt, "docker-bake.hcl")
+		require.NoError(t, err)
+		require.Equal(t, 1, len(c.Targets))
+		require.Equal(t, "3", *c.Targets[0].Args["foo"])
+
+		t.Setenv("FOO_JSON", "[10,11,12]")
+		_, err = ParseFile(dt, "docker-bake.hcl")
+		require.ErrorContains(t, err, "from CSV")
+	})
+
+	t.Run("override-ish variable has own JSON override", func(t *testing.T) {
+		dt := []byte(`
+            variable "FOO_JSON" {
+                type = list(number)
+            }
+
+            target "default" {
+                args = {
+                    foo = length(FOO_JSON)
+                }
+            }`)
+		t.Setenv("FOO_JSON_JSON", "[4,5,6]")
+		c, err := ParseFile(dt, "docker-bake.hcl")
+		require.NoError(t, err)
+		require.Equal(t, 1, len(c.Targets))
+		require.Equal(t, "3", *c.Targets[0].Args["foo"])
+	})
+
+	t.Run("JSON override trumps CSV when no var name conflict", func(t *testing.T) {
+		dt := []byte(`
+            variable "FOO" {
+                type = list(number)
+            }
+            
+            target "default" {
+                args = {
+                    foo = length(FOO)
+                }
+            }`)
+		t.Setenv("FOO", "1,2")
+		t.Setenv("FOO_JSON", "[3,4,5]")
+		c, err := ParseFile(dt, "docker-bake.hcl")
+		require.NoError(t, err)
+		require.Equal(t, 1, len(c.Targets))
+		require.Equal(t, "3", *c.Targets[0].Args["foo"])
+	})
+
+	t.Run("JSON override works with lowercase vars", func(t *testing.T) {
+		dt := []byte(`
+            variable "foo" {
+                type = number
+            }
+            
+            target "default" {
+                args = {
+                    bar = foo
+                }
+            }`)
+		// may seem reasonable, but not supported
+		t.Setenv("foo_json", "9000")
+		c, err := ParseFile(dt, "docker-bake.hcl")
+		require.NoError(t, err)
+		require.Equal(t, 1, len(c.Targets))
+		// a variable with no value has always resulted in an empty string
+		require.Equal(t, "", *c.Targets[0].Args["bar"])
+
+		t.Setenv("foo_JSON", "42")
```

Comment:
Are there implications for Windows here (whether the test itself, or the functionality)?

Reply:
Windows env are case insensitive. It is fine if `_json` works on windows but doesn't in unix.